### PR TITLE
Release/1.10.0

### PR DESCRIPTION
**Pull request by jeicher**

### CODE REVIEW

File: seq2simulate/revolver/models.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+
+
+class ValidationMixin(object):
+
+    def _is_valid_dna(self, seq):
+        valids = 'ATGCatgc-'
+        for i in set(seq):
+            if i not in valids: 
+                return False
+        return True, ''
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: why is there a string at the end?

File: seq2simulate/revolver/models.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+        self._codons = codons
+
+    def add_codon(self, codon):
+        if isinstance(codon, Codon):
+            self._codons.append(codon)
+        elif isinstance(codon, str):
+            self._codons.append(Codon(seq=codon))
+            
+    def del_codon(self, codon_i):
+        self._codons = [codon for i, codon in enumerate(self._codons) if i != codon_i]
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: this is the same as
```
del self._codons[codon_i]
```
not that important but

File: seq2simulate/revolver/models.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+                if mutation[0] == 'deletion':
+                    mutations.append('del{}{}'.format(self.loc_aa, convert_dna_to_aa(mutation[1])))
+                elif mutation[0] == 'insertion':
+                    mutations.append('ins{}{}'.format(self.loc_aa, convert_dna_to_aa(mutation[1])))
+                else:
+                    aa0 = convert_dna_to_aa(mutation[0])
+                    aa1 = convert_dna_to_aa(mutation[1])
+                    if aa0 != aa1:
+                        mutations.append('{}{}{}'.format(aa0, self.loc_aa, aa1))
+        except:
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: catch all excepts can cause insidious problems

File: seq2simulate/revolver/models.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+    Returns instantaneous rate of mutation using the simplified Goldman model in Nielsen (1998).
+    """
+    if len(codon1) != 3:
+        raise ValueError('codon must be a three-letter str')
+    if len(codon2) != 3:
+        raise ValueError('codon must be a three-letter str')
+
+    cat = mutation_category(codon1, codon2, codon_table=codon_table)
+    rate = 0
+    if cat == 'multisite':
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: mutation category returns ['multisite'] not 'multisite'

File: seq2simulate/revolver/models.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        rate = 0
+    if cat == ['synonymous', 'transversion']:
+        rate = codon_freq[codon2]
+    if cat == ['synonymous', 'transition']:
+        rate = kappa * codon_freq[codon2]
+    if cat == ['nonsynonymous', 'transversion']:
+        rate = omega * codon_freq[codon2]
+    if cat == ['nonsynonymous', 'transition']:
+        rate = omega * kappa * codon_freq[codon2]
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: change to elif's and then have an else that throws an exception if nothing matches.

File: seq2simulate/revolver/models.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+    Args:
+        q: Q matrix (generated by goldman_Q())
+        t: time or branch length over which to plot(if q was scaled, this is
+        the expected number of nucleotide mutations per codon)
+        codon: codon triplet to plot
+        logscale: log-scaled axes
+         
+    """
+    if not _CAN_PLOT:
+        print()
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: if not can plot we should prrobably return? so the following stuff doesn't fail?

File: seq2simulate/revolver/models.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+def choose_random_codon(codon_freq=None):
+    """
+    Args:
+        codon_freq: codon frequence dictionary
+    """
+    if codon_freq is None:
+        codon_freq = FEqual()
+
+    codons = [i for i in codon_freq.keys()]
```
Comment (MrHamdulay) [inline seq2simulate/revolver/models.py]: [style] equivalent to
```python
codons = codon_freq.keys()
```
I guess in python3 it returns an iterator...
list(codon_freq.keys()) is equivalent in that case

File: seq2simulate/revolver/trees.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+def get_list_of_tree_nodes(tree):
+    """
+    Perform a preorder traversal of 'tree', and return a list of all nodes.
+    """
+
+    def get_nodes(tree, nodes=[]):
+
+        if tree:
+            nodes.append(tree)
```
Comment (MrHamdulay) [inline seq2simulate/revolver/trees.py:154]: very cool

File: seq2simulate/revolver/trees.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+            get_nodes(tree.left, nodes=nodes)
+            get_nodes(tree.right, nodes=nodes)
+
+        return nodes
+
+    nodes = []
+    return get_nodes(tree, nodes=nodes)
+
+
+if __name__ == '__main__':
```
Comment (MrHamdulay) [inline seq2simulate/revolver/trees.py]: not needed

File: seq2simulate/revolver/trees.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
+import random
```
Comment (MrHamdulay) [inline seq2simulate/revolver/trees.py:1]: this stuff is cool.

**Review by MrHamdulay — APPROVED** [review]
this is awesome, very very minor mostly style comments.

File: seq2simulate/revolver/models.py (hunk 7 above)
Comment (jeicher) [reply seq2simulate/revolver/models.py]: yar, still unlearning Python 3

**Pull request merged**